File: algorithms/sort/merge.cpp
#include<bits/stdc++.h>
using namespace std;

void merge(vector<int> &arr, int low, int mid, int high){
  int a[high-low+1];
  int l = low, h = mid+1, k = 0;
  
  while(l <= mid && h <= high){
    if(arr[l] <= arr[h]) a[k++] = arr[l++];
    else a[k++] = arr[h++];
  }

  while(l <= mid) a[k++] = arr[l++];
  while(h <= high) a[k++] = arr[h++];

  for(int i = 0; i < k; i++) arr[low + i] = a[i];
}

void merge_sort(vector<int> &arr, int low, int high){
  if(low >= high) return;
  int mid = (low + high)/2;
  merge_sort(arr, low, mid);
  merge_sort(arr, mid+1, high);
  merge(arr, low, mid, high);
}

int main(){
  vector<int> arr = {4, 5, 1, 3, 2};
  merge_sort(arr, 0, 4);
  for (int i = 0; i < arr.size(); i++){
    cout << arr[i] << " ";
  }  
} 

File: algorithms/sort/merge.py
def merge(list, left, mid, right):
  l = left
  r = mid + 1
  merged_list = []

  while l <= mid and r <= right:
    if list[l] < list[r]:
      merged_list.append(list[l])
      l += 1
    else:
      merged_list.append(list[r])
      r += 1

  while l <= mid:
    merged_list.append(list[l])
    l += 1

  while r <= right:
    merged_list.append(list[r])
    r += 1

  for i in range(right - left + 1):
    list[left + i] = merged_list[i]

def merge_sort(list, left, right):
  if left < right:
    mid = left + (right - left)//2
    merge_sort(list, left, mid)
    merge_sort(list, mid + 1, right)
    merge(list, left, mid, right)

list = [4, 5, 9, 8, 7, 3, 2, 1, 6]
merge_sort(list, 0, len(list)-1)
print(list)
